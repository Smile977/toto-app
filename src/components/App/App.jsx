import React, { useState } from 'react';

import TodoContext from '../../context/context';

import { Header } from '../Header';
import { TodoList } from '../TodoList';

export const App = () => {
  const [text, setText] = useState('');
  const [todos, setTodos] = useState([
    {id: 0, title: 'First todo First todo First todo First todo First todo First todo First todo First todo First todo First todo First todo First todo First todo First todo', complited: false},
    {id: 1, title: 'Second todo', complited: true},
    {id: 2, title: 'One more todo', complited: true},
    {id: 3, title: 'The next todo', complited: false},
    {id: 4, title: 'Last todo', complited: false},
  ]);

  const addTodo = (value) => {
    const newId = todos.length ? todos[todos.length - 1].id : 0;    
    setTodos([...todos, {id: newId + 1, title: value, complited: false}]);    
  }

  return (
    <>
      <TodoContext.Provider value={{text, setText, addTodo, todos, setTodos}}>
        <Header />
        <TodoList />
      </TodoContext.Provider> 
    </>
  );
}
