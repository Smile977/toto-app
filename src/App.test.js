import React from "react";
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { App } from "./components/App";
import { render, waitFor, screen } from '@testing-library/react';
import userEvent from "@testing-library/user-event";

const server = setupServer(
  rest.get('/todos', (req, res, ctx) => {
    return res(ctx.json([
        {
          id: 'df1ce37b-c137-4adb-be9f-6195f9b86b20',
          title: 'Second todo',
          complited: true
        },
        {
          id: 'a52543fb-9fcb-4324-af8b-ed1970670122',
          title: 'One more todo',
          complited: true
        }      
      ]))
  }),
  rest.post('/todos/add', (req, res, ctx) => {
    return res(ctx.json([
        {
          id: 'ee492d70-fe2f-41b5-a54f-3906d61f6419',
          title: 'new todo',
          complited: false
        }        
      ]))
  })
)

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('render App', () => {
  it('before loading data from server', () => {
    
    const { asFragment } = render(<App />);
    expect(asFragment(<App />)).toMatchSnapshot();
  })

  it('get title not totost', () => {
    render(<App />);

    const element = screen.getByTestId('title-not-todos');
  })


  it('Loads and todos', async () => {
    render(<App url="/todos" />);

    await waitFor(() => screen.getByText(/todos:/i));
    expect(await screen.findByText('Second todo')).toBeInTheDocument();
  }) 

  it('Add new todo', async () => {
    render(<App url="/todos/add" />);     
    
    userEvent.type(screen.getByRole('textbox'), 'new todo');
    userEvent.type(screen.getByRole('textbox'), '{enter}');          
    expect(await screen.findByText(/new todo/i)).toBeInTheDocument();
  })

  it('change checkbox', async () => {
    render(<App />);
    userEvent.type(screen.getByRole('textbox'), 'Hello, World!');
    expect(await screen.getByRole('textbox').value).toBe('Hello, World!');
  })  
});
