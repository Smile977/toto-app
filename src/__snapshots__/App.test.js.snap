// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render App render App 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="header-container"
    >
      <h3
        class="header-container__title"
      >
        Todo List
      </h3>
      <div
        class="function search() { [native code] }"
      >
        <div
          class="search-container"
        >
          <form
            class="search-container__form"
          >
            <input
              class="search-container__form-input"
              placeholder="add new todo"
              type="text"
              value=""
            />
          </form>
        </div>
      </div>
    </div>
  </header>
  <div
    class="todos"
  >
    <h3
      class="todos-title"
    >
      There are haven't todos yet
    </h3>
  </div>
</DocumentFragment>
`;
